src/files/gcp/GcpStore.ts

```diff
@@ -0,0 +1,111 @@
+import { Storage, Bucket } from '@google-cloud/storage';
+import * as debug from 'debug';
+
+import * as config from '../../config';
+
+const d = debug('nucleus:gcp');
+
+export default class GcpStore implements IFileStore {
+  private readonly bucket: Bucket;
+  private readonly publicUrl: string;
+  private readonly cachePolicyTransient: string;
+  constructor() {
+    this.cachePolicyTransient = config.gcp.cachePolicyTransient;
+    this.bucket = new Storage({
+      projectId: config.gcp.projectId,
+      autoRetry: true,
+    }).bucket(config.gcp.bucketName);
+
+    this.publicUrl = config.gcp.publicUrl
+      ? config.gcp.publicUrl
+      : `https://storage.googleapis.com/${config.gcp.bucketName}`;
+  }
+
+  public async hasFile(key: string) {
+    const [exists] = await this.bucket.file(key).exists();
+    return exists;
+  }
+
+  public async getFileSize(key: string) {
+    try {
+      const [metadata] = await this.bucket.file(key).getMetadata();
+      return metadata.size;
+    } catch (err) {
+      d(
+        `failed to fetch file size for '${key}', returning 0 instead. Error message: ${
+          err.message
+        }`,
+      );
+      return 0;
+    }
+  }
+
+  public async putFile(key: string, data: Buffer, overwrite = false) {
+    d(`Putting file: '${key}', overwrite=${overwrite ? 'true' : 'false'}`);
+    const keyExists = async () => await this.hasFile(key);
+    let wrote = false;
+    if (overwrite || !(await keyExists())) {
+      d(
+        "Deciding to write file (either because overwrite is enabled or the key didn't exist)",
+      );
+
+      await new Promise((resolve, reject) => {
```

Comment:
Can we reduce the scope of this promise to just the stream logic.

I.e. move this line down to line 68 (just before `stream.on('finish')`)

Reply:
done